sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -150,6 +152,10 @@ case class InsertIntoHadoopFsRelationCommand(
         }
       }
 
+      val partitionSet = AttributeSet(partitionColumns)
+      val dataColumns = query.output.filterNot(partitionSet.contains)
```

Comment:
We should use `outputColumns` instead of `query.output`, cc @gengliangwang

Reply:
+1, it should be `outputColumns ` here, which is the output columns of analyzed plan. See #20020 for details.